File: trace/scene.cc
#include "trace/scene.h"

#include <glm/gtc/epsilon.hpp>
#include <map>

#include "geometry/triangle.h"
#include "kdtree/build.h"
#include "kdtree/kdtree.h"
#include "trace/camera.h"
#include "trace/material.h"
#include "trace/mcsampling.h"
#include "wavefront/mtl.h"
#include "wavefront/obj.h"

using glm::vec3;
using std::map;
using std::string;
using std::vector;

namespace trace {
namespace {
constexpr float EPSILON = 0.0001f;

Material* blend1_from_wavefront(const wavefront::Material& material) {
  if (glm::epsilonEqual(material.transparency, 1.0f, EPSILON)) {
    return new SpecularRefractionMaterial(material.ior);
  } else if (glm::epsilonEqual(material.transparency, 0.0f, EPSILON)) {
    return new DiffuseMaterial(material.diffuse);
  } else {
    return new BlendMaterial(new SpecularRefractionMaterial(material.ior),
                             new DiffuseMaterial(material.diffuse),
                             material.transparency);
  }
}

Material* blend0_from_wavefront(const wavefront::Material& material,
                                Material* blend1) {
  if (glm::epsilonEqual(material.refl90, 1.0f, EPSILON)) {
    return new FresnelBlendMaterial(
        new SpecularReflectionMaterial(material.specular), blend1,
        material.refl0);
  } else if (glm::epsilonEqual(material.refl90, 0.0f, EPSILON)) {
    return blend1;
  } else {
    return new BlendMaterial(
        new FresnelBlendMaterial(
            new SpecularReflectionMaterial(material.specular), blend1,
            material.refl0),
        blend1, material.refl90);
  }
}

Material* material_from_wavefront(const wavefront::Material& material) {
  return blend0_from_wavefront(material, blend1_from_wavefront(material));
}

void update_pointer_to_material(
    const std::map<std::string, Material*>& materials,
    std::vector<geometry::Triangle>* triangles) {
  for (geometry::Triangle& triangle : *triangles) {
    triangle.tag = materials.at(*static_cast<const std::string*>(triangle.tag));
  }
}
}  // namespace

vector<SphereLight> lights_from_mtl(const wavefront::Mtl& mtl) {
  vector<SphereLight> lights;
  for (const wavefront::Light& light : mtl.lights) {
    lights.emplace_back(light.center, light.color, light.intensity,
                        light.radius);
  }
  return lights;
}

vector<Camera> cameras_from_mtl(const wavefront::Mtl& mtl) {
  vector<Camera> cameras;
  for (const wavefront::Camera& camera : mtl.cameras) {
    cameras.emplace_back(camera.position, camera.target, camera.up,
                         glm::radians(camera.fov));
  }
  return cameras;
}

map<string, Material*> materials_from_mtl(const wavefront::Mtl& mtl) {
  map<string, Material*> materials;
  for (const wavefront::Material& material : mtl.materials) {
    materials[material.name] = material_from_wavefront(material);
  }
  return materials;
}

vector<geometry::Triangle> triangles_from_obj(const wavefront::Obj& obj) {
  vector<geometry::Triangle> triangles;
  for (const wavefront::Chunk& chunk : obj.chunks) {
    for (const wavefront::Face& polygon : chunk.polygons) {
      // TODO: ensure the lifetime of somehow &chunk.material
      triangles.push_back(
          {IndexVertex(obj, polygon.p1.v), IndexVertex(obj, polygon.p2.v),
           IndexVertex(obj, polygon.p3.v), IndexNormal(obj, polygon.p1.n),
           IndexNormal(obj, polygon.p2.n), IndexNormal(obj, polygon.p3.n),
           IndexTexcoord(obj, polygon.p1.t), IndexTexcoord(obj, polygon.p2.t),
           IndexTexcoord(obj, polygon.p3.t), &chunk.material});
    }
  }
  return triangles;
}

Scene::Scene(const wavefront::Obj& obj, const wavefront::Mtl& mtl)
    : triangles_(triangles_from_obj(obj)),
      materials_(materials_from_mtl(mtl)),
      cameras_(cameras_from_mtl(mtl)),
      lights_(lights_from_mtl(mtl)),
      kdtree_(kdtree::build(triangles_)) {
  update_pointer_to_material(materials_, &triangles_);
}
}  // namespace trace

File: src/scene.rs
use geometry::triangle::Triangle;
use nalgebra::{UnitVector3, Vector2, Vector3};
use std::{collections::BTreeMap, sync::Arc};
use wavefront::{mtl, obj};

use crate::{
    camera::Camera,
    light::SphericalLight,
    material::{
        BlendMaterial, DiffuseReflectiveMaterial, FresnelBlendMaterial, Material,
        SpecularReflectiveMaterial, SpecularRefractiveMaterial,
    },
};

#[derive(Clone, Copy, Debug, PartialEq)]
pub struct TriangleNormals {
    pub n0: Vector3<f32>,
    pub n1: Vector3<f32>,
    pub n2: Vector3<f32>,
}

impl TriangleNormals {
    pub fn lerp(&self, u: f32, v: f32) -> UnitVector3<f32> {
        UnitVector3::new_normalize((1.0 - (u + v)) * self.n0 + u * self.n1 + v * self.n2)
    }
}

#[derive(Clone, Copy, Debug, PartialEq)]
pub struct TriangleTexcoords {
    pub uv0: Vector2<f32>,
    pub uv1: Vector2<f32>,
    pub uv2: Vector2<f32>,
}

pub struct TriangleData {
    pub triangle: Triangle,
    pub normals: TriangleNormals,
    pub texcoords: TriangleTexcoords,
    pub material: Arc<dyn Material + Send + Sync>,
}

pub struct Scene {
    pub triangle_data: Vec<TriangleData>,
    pub cameras: Vec<Camera>,
    pub lights: Vec<SphericalLight>,
    pub environment: Vector3<f32>,
}

fn collect_triangle_data(obj: &obj::Obj, mtl: &mtl::Mtl) -> Vec<TriangleData> {
    let materials = materials_from_mtl(mtl);
    obj.chunks
        .iter()
        .flat_map(|chunk| {
            chunk.faces.iter().map(|face| {
                let triangle = Triangle {
                    v0: obj.index_vertex(&face.p0).into(),
                    v1: obj.index_vertex(&face.p1).into(),
                    v2: obj.index_vertex(&face.p2).into(),
                };
                let normals = TriangleNormals {
                    n0: obj.index_normal(&face.p0).into(),
                    n1: obj.index_normal(&face.p1).into(),
                    n2: obj.index_normal(&face.p2).into(),
                };
                let texcoords = TriangleTexcoords {
                    uv0: obj.index_texcoord(&face.p0).into(),
                    uv1: obj.index_texcoord(&face.p1).into(),
                    uv2: obj.index_texcoord(&face.p2).into(),
                };
                TriangleData {
                    triangle,
                    normals,
                    texcoords,
                    material: materials[chunk.material.as_str()].clone(),
                }
            })
        })
        .collect::<Vec<_>>()
}

fn blend_from_mtl(material: &mtl::Material) -> Arc<dyn Material + Send + Sync> {
    let reflection = DiffuseReflectiveMaterial {
        reflectance: material.diffuse_reflection.into(),
    };
    let refraction = SpecularRefractiveMaterial {
        index_of_refraction: material.index_of_refraction,
    };
    let specular = SpecularReflectiveMaterial {
        reflectance: material.specular_reflection.into(),
    };
    let transparency_blend = BlendMaterial {
        first: refraction,
        second: reflection,
        factor: material.transparency,
    };
    let fresnel_blend = FresnelBlendMaterial {
        reflection: specular,
        refraction: transparency_blend.clone(),
        r0: material.reflection_0_degrees,
    };
    Arc::new(BlendMaterial {
        first: fresnel_blend,
        second: transparency_blend,
        factor: material.reflection_90_degrees,
    })
}

fn materials_from_mtl(mtl: &mtl::Mtl) -> BTreeMap<&str, Arc<dyn Material + Send + Sync>> {
    mtl.materials
        .iter()
        .map(|m| (m.name.as_str(), blend_from_mtl(m)))
        .collect::<BTreeMap<_, _>>()
}

fn collect_cameras(mtl: &mtl::Mtl) -> Vec<Camera> {
    mtl.cameras
        .iter()
        .map(|camera| {
            Camera::new(
                &camera.position.into(),
                &camera.target.into(),
                &camera.up.into(),
                camera.fov,
            )
        })
        .collect()
}

fn collect_lights(mtl: &mtl::Mtl) -> Vec<SphericalLight> {
    mtl.lights
        .iter()
        .map(|light| {
            SphericalLight::new(
                light.position.into(),
                light.radius,
                &light.color.into(),
                light.intensity,
            )
        })
        .collect()
}

impl Scene {
    pub fn from_wavefront(obj: &obj::Obj, mtl: &mtl::Mtl) -> Scene {
        Scene {
            triangle_data: collect_triangle_data(obj, mtl),
            cameras: collect_cameras(mtl),
            lights: collect_lights(mtl),
            environment: Vector3::new(0.8, 0.8, 0.8),
        }
    }
}
